rollup-node.md

```diff
@@ -0,0 +1,296 @@
+driver]. This document is currently only concerned with the specification of the
+rollup driver.
+
+## Table of Contents
+
+- [L2 Chain Derivation][l2-chain-derivation]
+  - [From L2 derivation inputs to payload attributes][payload-attr]
+  - [Payload Transaction Format][payload-format]
+  - [Building the L2 block with the execution engine][calling-exec-engine]
+- [Handling L1 Re-Orgs][l1-reorgs]
+- [Finalization Guarantees][finalization]
+
+## L2 Chain Derivation
+[l2-chain-derivation]: #l2-chain-derivation
+
+This section specifies how the [rollup driver] derives one L2 block from every
+L1 block. The L2 block will carry the L1 block attributes (as a *[L1 attributes
+transaction]*) well as all L2 transactions deposited by users in the L1 block
+(*[deposits]*).
+
+[L1 attributes transaction]: /glossary.md#l1-attributes-transaction
+
+### From L2 derivation inputs to payload attributes
+[payload-attr]: #From-L2-derivation-inputs-to-payload-attributes
+
+[`PayloadAttributesOPV1`]: #From-L2-derivation-inputs-to-payload-attributes
+
+The rollup reads the following [L2 derivation inputs] for each L1 block:
+
+- L1 block attributes
+   - block number
+   - timestamp
+   - basefee
+   - *random* (the output of the [`RANDOM` opcode][random])
+- [deposits]
+
+[random]: https://eips.ethereum.org/EIPS/eip-4399
+
+A deposit is an L2 transaction that has been submitted on L1, via a transaction
+sent to the [deposit feed contract][deposit-feed].
+
+While deposits are notably (but not only) used to "deposit" (bridge) ETH and
+tokens to L2, the word *deposit* should be understood as "a transaction
+*deposited* to L2".
+
+The L1 attributes are read from the L1 block header, while deposits are read
+from the block's log entries. Refer to the [**deposit feed contract
+specification**][deposit-feed] for details on how deposits are encoded as log
+entries.
+
+[deposit-feed-spec]: TODO
+
+> **TODO LINK** deposit feed contract specification
+
+From these L2 derivation inputs, the rollup node constructs an expanded version
+of the [Engine API PayloadAttributesV1 object][PayloadAttributesV1]:
+
+[PayloadAttributesV1]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#payloadattributesv1
+
+```
+PayloadAttributesOPV1: {
+	timestamp: QUANTITY
+	random: DATA (32 bytes)
+	suggestedFeeRecipient: DATA (20 bytes)
+	transactions: array of DATA
+}
+```
+
+The type notation used here refers to the [HEX value encoding] used by the
+[Ethereum JSON-RPC API specification][JSON-RPC-API], as this structure will need to
+be sent over JSON-RPC. `array` refers to a JSON array.
+
+[HEX value encoding]: https://eth.wiki/json-rpc/API#hex-value-encoding
+[JSON-RPC-API]: https://eth.wiki/json-rpc/API
+
+The object properties must be set as follows:
+
+- `timestamp` is set to the current [unix time] (number of elapsed seconds since
+  00:00:00 UTC on 1 January 1970), rounded to the closest multiple of 2 seconds.
```

Comment:
```suggestion
- `timestamp` is set to the current [unix time], rounded to the closest multiple of 2 seconds.
```

I think this is a case where we can just link out to the reference without adding context.

Reply:
I think its worth also indicating that the timestamps must be unique, ie. no two L2 blocks have the same one.

Reply:
Which reference would you link though? We're not copying this from L1 (1)! This is also different from the L1 spec (which doesn't round at 2s), and that spec is hard to link (because it's inside a pdf).

Also even if we reused the spec AND it was reusable, this is literally 10 words, redundancy seems worth it here.

(1) Although maybe we should at this stage - good question! In the context of future stages, do we want deposit timestamp == L1 timestamp or do we want to reflect our potential processing lag in the timestamp (imagining we go with the 1 block / 2s proposal of Karl)?